server-common/src/main/java/org/apache/kafka/server/common/Feature.java

```diff
@@ -0,0 +1,349 @@
+                feature.name(), latestProduction.featureLevel(), defaultVersion.featureLevel()));
+        }
+
+        for (Map.Entry<String, Short> dependency: latestProduction.dependencies().entrySet()) {
+            String dependencyFeatureName = dependency.getKey();
+            if (!dependencyFeatureName.equals(MetadataVersion.FEATURE_NAME)) {
+                Feature dependencyFeature = featureFromName(dependencyFeatureName);
+                if (!dependencyFeature.isProductionReady(dependency.getValue())) {
+                    throw new IllegalArgumentException(String.format("Latest production FeatureVersion %s=%s " +
+                            "has a dependency %s=%s that is not production ready. (%s latest production: %s)",
+                        feature.name(), latestProduction.featureLevel(), dependencyFeature.name(), dependency.getValue(),
+                        dependencyFeature.name(), dependencyFeature.latestProduction()));
+                }
+            } else {
+                if (dependency.getValue() > MetadataVersion.LATEST_PRODUCTION.featureLevel()) {
+                    throw new IllegalArgumentException(String.format("Latest production FeatureVersion %s=%s " +
+                            "has a dependency %s=%s that is not production ready. (%s latest production: %s)",
+                        feature.name(), latestProduction.featureLevel(), MetadataVersion.FEATURE_NAME, dependency.getValue(),
+                        MetadataVersion.FEATURE_NAME, MetadataVersion.LATEST_PRODUCTION.featureLevel()));
+                }
+            }
+        }
+
+        for (MetadataVersion metadataVersion: MetadataVersion.values()) {
+            // Only checking the kraft metadata versions.
+            if (metadataVersion.compareTo(MetadataVersion.MINIMUM_KRAFT_VERSION) < 0) {
+                continue;
+            }
+
+            defaultVersion = feature.defaultVersion(metadataVersion);
+            for (Map.Entry<String, Short> dependency: defaultVersion.dependencies().entrySet()) {
+                String dependencyFeatureName = dependency.getKey();
+                if (!dependencyFeatureName.equals(MetadataVersion.FEATURE_NAME)) {
+                    Feature dependencyFeature = featureFromName(dependencyFeatureName);
+                    if (dependency.getValue() > dependencyFeature.defaultLevel(MetadataVersion.LATEST_PRODUCTION)) {
+                        throw new IllegalArgumentException(String.format("Default FeatureVersion %s=%s has " +
+                                "a dependency %s=%s that is ahead of its default value %s.",
+                            feature.name(), defaultVersion.featureLevel(), dependencyFeature.name(), dependency.getValue(),
+                            dependencyFeature.defaultLevel(MetadataVersion.LATEST_PRODUCTION)));
+                    }
+                } else {
+                    if (dependency.getValue() > defaultVersion.bootstrapMetadataVersion().featureLevel()) {
```

Comment:
The two are not necessarily equal. For instance, when `metadataVersion=120`, the feature has versions
- FV_1, bootstrap MV = 119
- FV_2, bootstrap MV = 121

Then `defaultVersion.bootstrapMetadataVersion()` will be 119. Using `metadataVersion` would also work since we do check when the two are equal, but I feel comparing with the bootstrap MV makes more sense.